diy-omd.ino

```diff
@@ -85,6 +106,244 @@ void loop() {
+    motorState = GO;
+    secondaryTimeMillis = 0;
+  } else if (MODE == AXE_550_CALIBRATE) {
+    // Serial.println("AXE_550_calibrate init");
+    // Serial.println(secondaryTimeMillis);
+    digitalWrite(13, secondaryTimeMillis % 2000 > 1000 ? HIGH : LOW);
+
+    if (secondaryTimeMillis > 10000) {
+      motorState = GO;
+      secondaryTimeMillis = 0;
+    }
+  }
+}
+
+void go() {
+  digitalWrite(13, HIGH);
+  ESC.write(goValue);
+  if (secondaryTimeMillis > onSequence && MODE == OMD) {
+    //Serial.println("go done");
+    motorState = STOP;
+    secondaryTimeMillis = 0;
+  } else if (MODE == BALANCE) {
+    // do nothing: we just keep going in balance mode!
+  } else if (MODE == AXE_550_CALIBRATE) {
+    // Serial.println("AXE_550_calibrate go");
+    // Serial.println(secondaryTimeMillis % 1000);
+    digitalWrite(13, secondaryTimeMillis % 1000 > 500 ? HIGH : LOW);
+
+    if (secondaryTimeMillis > 10000) {
+      motorState = REVERSE;
+      secondaryTimeMillis = 0;
+    }
+  }
+}
+
+void processMode(const char* modeString) {
+  int mode = atoi(kModeToCode[modeString]);
+
+  Serial.println("Switching mode to");
+  switch (mode) {
+    case OMD: {
+      Serial.println("OMD");
+      MODE = OMD;
+      secondaryTimeMillis = 0;
+      break;
+    }
+
+    case BALANCE: {
+      Serial.println("BALANCE");
+      MODE = BALANCE;
+      break;
+    }
+
+    case AXE_550_CALIBRATE: {
+      Serial.println("CALIBRATE");
+      MODE = AXE_550_CALIBRATE;
+      break;
+    }
+
+    case OFF: {
+      Serial.println("OFF");
+      MODE = OFF;
+      break;
+    }
+  }
+
+  int tempGoValue = MODE == OMD ? OMD_goValue : (
+    MODE == BALANCE ? BALANCE_goValue : AXE_550_CALIBRATE_goValue
+  );
+
+  if (tempGoValue <= MAX_GO_VALUE) {
+    goValue = MODE == OMD ? OMD_goValue : (
+      MODE == BALANCE ? BALANCE_goValue : AXE_550_CALIBRATE_goValue
+    );
+  }
+}
+
+void processIncomingBTData() {
+  char appData;
+
+  if (btSerial.available() > 0) {
+    appData = btSerial.read();
+
+    if (appData == PARSE_START || start == true) {
+      inData += appData;  // save the data in string format
+      start = true;
+    }
+    //Serial.println(appData);
+  }
+
+  if (appData == PARSE_END) {
+    start = false;
+    processJSON();
+  }
+}
+
+void processJSON() {
+  Serial.println("Processing JSON...");
+  inData.replace(">", "");
+  inData.replace("<", "");
+  Serial.println("Parsing the follow: ");
+  Serial.println(inData);
+
+  // Deserialize the JSON document
+  char buffer[50];
+  inData.toCharArray(buffer, 50);
+  DeserializationError error = deserializeJson(doc, buffer);
+  inData = "";
+
+  // Test if parsing succeeds.
+  if (error) {
+    Serial.print(F("deserializeJson() failed: "));
+    Serial.println(error.f_str());
+    return;
+  }
+
+  processCmd();
+}
+
+void processCmd() {
+  const char* modeString = doc["mode"];
+  const char* on = doc["on"];
+  const char* off = doc["off"];
+  const char* speed = doc["hz"];
```

Comment:
Calling this `hz` in the JSON was pretty confusing for me to read here, since that seems to indicate we have exact frequency control. Instead, I think this is just passed in directly as the variable we send to `ESC.writeValue`, yeah? Can we make it something like `escSpeed` instead?

Reply:
Ok, i'll change it. It's just that I wasn't able to send long key/value pairs from the UI to Arduino via Bluetooth. It would restart the Bluetooth pairing and get all messed up.